onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlDFT.h

```diff
@@ -0,0 +1,695 @@
+
+    struct LoopRange
+    {
+        unsigned Left;
+        unsigned Right;
+        unsigned End;
+        unsigned CalculateIndex(unsigned index) {
+            if (index > 0 && index < End)
+            {
+                unsigned range = Right - Left + 1;
+                index = Left + (index - 1) % range;
+            }
+            else if (index == End) {
+                index = Right + 1;
+            }
+            return index;
+        }
+    };
+    LoopRange loopRange_ = {};
+
+    struct DFTShaderConstants
+    {
+        uint32_t StartIndex;
+        uint32_t ElementCount;
+        uint32_t DFTIteration;
+        uint32_t IsInverse;
+        uint32_t InputSizes[4];
+        uint32_t InputStrides[4];
+        uint32_t OutputSizes[4];
+        uint32_t OutputStrides[4];
+        float Scale;
+    };
+
+public:
+    GpuDFTOperator(IMLOperatorKernelCreationContext* context)
+    {
+        ComPtr<IUnknown> executionObject;
+        context->GetExecutionInterface(executionObject.GetAddressOf());
+
+        ComPtr<ID3D12GraphicsCommandList> commandList;
+        executionObject.As(&commandList);
+
+        ORT_THROW_IF_FAILED(commandList->GetDevice(IID_ID3D12Device, &m_device));
+
+
+        ORT_THROW_IF_FAILED(context->GetAttribute("axis", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&m_axis)));
+
+        int64_t isInverseInt;
+        ORT_THROW_IF_FAILED(context->GetAttribute("inverse", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&isInverseInt)));
+        m_isInverse = static_cast<bool>(isInverseInt);
+
+        int64_t isOnesidedInt;
+        ORT_THROW_IF_FAILED(context->GetAttribute("onesided", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&isOnesidedInt)));
+        m_isOnesided = static_cast<bool>(isOnesidedInt);
+
+        ComPtr<IMLOperatorTensorShapeDescription> shapeDesc;
+        context->GetTensorShapeDescription(shapeDesc.GetAddressOf());
+
+        // Get the input and output shape sizes
+        uint32_t inputDimsSize;
+        ORT_THROW_IF_FAILED(shapeDesc->GetInputTensorDimensionCount(0, &inputDimsSize));
+        uint32_t outputDimsSize;
+        ORT_THROW_IF_FAILED(shapeDesc->GetOutputTensorDimensionCount(0, &outputDimsSize));
+        ORT_THROW_HR_IF(E_FAIL, inputDimsSize != outputDimsSize);
+
+        // Get the input shape
+        m_inputDims.resize(inputDimsSize);
+        ORT_THROW_IF_FAILED(shapeDesc->GetInputTensorShape(0, static_cast<uint32_t>(m_inputDims.size()), m_inputDims.data()));
+
+        // Get the output shape
+        m_outputDims.resize(outputDimsSize);
+        ORT_THROW_IF_FAILED(shapeDesc->GetOutputTensorShape(0, static_cast<uint32_t>(m_outputDims.size()), m_outputDims.data()));
+
+        // For the number of total elements in the input and output shapes
+        m_outputDataSize = std::accumulate(m_outputDims.begin(), m_outputDims.end(), 1, std::multiplies<uint32_t>());
```

Comment:
>std::accumulate(m_outputDims.begin(), m_outputDims.end(), 1, std::multiplies<uint32_t>()) [](http://example.com/codeflow?start=27&length=89)

[consider, but I don't care strongly]
There's a helper function in MLOperatorAuthorHelper.h

`= ComputeElementCountFromDimensions(m_outputDims)`